examples/image_ocr.py

```diff
@@ -0,0 +1,433 @@
+            self.cur_train_index += self.minibatch_size
+            if self.cur_train_index >= self.val_split:
+                self.cur_train_index = self.cur_train_index % 32
+                (self.X_text, self.Y_data, self.Y_len) = shuffle_mats_or_lists(
+                    [self.X_text, self.Y_data, self.Y_len], self.val_split)
+            yield ret
+
+    def next_val(self):
+        while 1:
+            ret = self.get_batch(self.cur_val_index, self.minibatch_size, train=False)
+            self.cur_val_index += self.minibatch_size
+            if self.cur_val_index >= self.num_words:
+                self.cur_val_index = self.val_split + self.cur_val_index % 32
+            yield ret
+
+    def on_train_begin(self, logs={}):
+        # translational invariance seems to be the hardest thing
+        # for the RNN to learn, so start with <= 4 letter words.
+        self.build_word_list(5, 1)
+
+    def on_epoch_begin(self, epoch, logs={}):
+        # After 10 epochs, translational invariance should be learned
+        # so start feeding longer words and eventually multiple words with spaces
+        if epoch == 10:
+            self.build_word_list(8, 1)
+        if epoch == 20:
+            self.build_word_list(8, 0.5)
+        if epoch == 30:
+            self.build_word_list(12, 0.5)
+
+# the actual loss calc occurs here despite it not being
+# a true Keras loss function
+def ctc_lambda_func(inputs):
+    # theano doesn't give the last operation a name for some reason,
+    # while tensorflow adds ":0" to names..so use a complicated cleanup func
+    inputs = {str(i.name).split(":")[0]: i for i in inputs}
+    # ..and some package specific code
+    if K._BACKEND == "tensorflow":
+        y_pred = inputs["div"]
+    else:
+        y_pred = inputs["None"]
+    # the 2 is critical here since the first couple outputs of the RNN
+    # tend to be garbage:
+    y_pred = y_pred[:, 2:, :]
+
+    return K.ctc_batch_cost(inputs["the_labels"], y_pred,
+                            inputs["input_length"], inputs["label_length"])
+
+# keras still wants a loss func in this particular parameter format
+def dummy_func(y_true_dummy, y_pred):
+    return y_pred
+
+# For a real OCR application, this should be beam search with a dictionary
+# and language model.  For this example, best path is sufficient.
+def decode_batch(test_func, word_batch):
+    import itertools
+    out = test_func([word_batch])[0]
+    ret = []
+    for j in range(out.shape[0]):
+        out_best = list(np.argmax(out[j, 2:], 1))
+        out_best = [k for k, g in itertools.groupby(out_best)]
+        # 26 is space, 27 is CTC blank char
+        outstr = ""
+        for c in out_best:
+            if c >= 0 and c < 26:
+                outstr += chr(c + ord('a'))
+            elif c == 26:
+                outstr += " "
+        ret.append(outstr)
+    return ret
+
+
+class VizCallback (keras.callbacks.Callback):
+
+    num_display_words = 6
```

Comment:
Better to make this a constructor argument than a class attribute. Class attributes use the same pointer across all class instances rather than a different pointer with every instance, so it's a major antipattern.